File: pset6/credit/credit.py
from cs50 import get_int
from sys import exit


def main():
    card_number = str(get_int("Number: "))
    result = checksum(card_number)

    if not result:
        print("INVALID")
        exit(0)

    first_digits = int("".join(card_number[0:2]))

    if len(card_number) == 15 and first_digits in [37, 37]:
        print("AMEX")

    elif len(card_number) == 16 and first_digits in [51, 52, 53, 54, 55]:
        print("MASTERCARD")

    elif len(card_number) in [13, 16] and str(first_digits)[0] == "4":
        print("VISA")

    exit(0)

def checksum(card_number):
    n = len(card_number) - 2
    first_sum = second_sum = 0

    for i in range(n, -1, -2):
        digits = str(int(card_number[i]) * 2)
        first_sum += sum([int(digit) for digit in digits])
        second_sum += int(card_number[i + 1])

        if i - 1 == 0 and n % 2 != 0:
            second_sum += int(card_number[0])

    return True if (first_sum + second_sum) % 10 == 0 else False

main()

File: pset1/credit/credit.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <cs50.h>


void card_name(long card_number);
bool checksum(long card_number);
int ascii_to_int(char c);

int main(void)

{

    long card_number = get_long("NUMBER: ");

    if (checksum(card_number))

    {
        card_name(card_number);
    }

    else

    {
        printf("INVALID\n");
    }
}




void card_name(long card_number)

{
    char number[255], digits[3];

    int lenght = sprintf(number, "%li", card_number);

    // join two digits

    digits[0] = number[0];
    digits[1] = number[1];
    digits[2] = '\0';

    if (lenght == 15 && (strncmp(digits, "34", 2) == 0 || strncmp(digits, "37", 2) == 0))

    {
        printf("AMEX\n");
    }

    else if (lenght == 16 && (strtol(digits, NULL, 0) >= 51 && strtol(digits, NULL, 0) <= 55))

    {
        printf("MASTERCARD\n");
    }

    else if ((lenght >= 13 && lenght <= 16) && strncmp(digits, "4", 1) == 0)

    {
        printf("VISA\n");
    }

    else

    {
        printf("INVALID\n");
    }
}


bool checksum(long card_number)

{
    int first_sum = 0, second_sum = 0;
    char number[255], n[5];

    int lenght = sprintf(number, "%li", card_number);

    for (int i = lenght - 2; i >= 0;)

    {

        sprintf(n, "%i", ascii_to_int(number[i])*2);
        if (strlen(n) > 1)

            {
                for (int j = 0; j < strlen(n); j++)

            {
                first_sum += ascii_to_int(n[j]);

            }

            }

            else
            {
                first_sum += ascii_to_int(number[i]) * 2;
            }

        second_sum +=  ascii_to_int(number[i + 1]);


        if (i == 1)
        {
            i--;
        }

        else

        {
            i -= 2;
        }

    }

    return (first_sum + second_sum) % 10 == 0;
}

int ascii_to_int(char c)

 {
     return c - 48;

}